### ISSUE TITLE
"!is" is treated as "is", need a compilation error and prompt change to 'is!'?

### ISSUE THREAD

**Issue by eggfly** [dart-lang/sdk#48496]

**Comment by lrhn** [comment]
Giving a warning sounds like a job for the analyzer.

If it doesn't already give a warning for `a! is int`, which it would if `a` is already non-nullable, it might want to consider the pattern as something that can be warned against in general.

(I don't see any good reason to do `!` before `is`. Usually the `is` promotes, and then you don't need the `!`. Even if not, it seems spurious to have a `!` test in the middle of something else.)

**Comment by srawlins** [comment]
I think this probably should be implemented as a lint rule, rather than a warning... but... I suppose it prevents bugs ("you didn't type what you think you typed"). I'll keep it as a Warning request.

**Comment by mateusfccp** [comment]
Why does the parser even allow it? Shouldn't we consider `id !` a parsing error?

**Comment by srawlins** [comment]
Broadly, the parser does not care about whitespace between tokens so `id!` is just as legal as `id !` (a null-assertion postfix operator). I'm not a parser expert, but I think making `id !` a parser error (as in `id !is int`) would introduce an odd inconsistency in the language.

**Comment by lrhn** [comment]
The parser being whitespace aware is not impossible, and it is done in a few places to avoid ambiguities. It better be worth it, though, and be something that you are very unlikely to hit by accident. 

In the case of a trailing `!`, something like `x!==2` will tokenize as `x != = 2`. People don't seem to have a problem avoiding that. 

If rather warm generally about `x!is T` than try to find a way for the parser to distinguish `x! is` from `x !is`. It's just not useful to check for `null` just before an `is` check.

**Comment by FMorschel** [comment]
I've made a CL for this https://dart-review.googlesource.com/c/sdk/+/449140

**Comment by bwilkerson** [comment]
I don't think we want to produce a warning for valid code. We already report the `!` in `nonNullableVariable!` as dead code, so this would only apply in cases where the null-check operator is valid.

I'm not convinced that we want to add a lint for this either because (a) there will likely be too many false positives and (b) the problem isn't likely to happen often enough to be worth the extra support burden. I might be wrong about one or either of those assumptions. I suppose a lint could trigger only when the `!` is next to the `is`, but that only works until the user formats their code, so that really doesn't seem very useful either.

**Comment by FMorschel** [comment]
> I suppose a lint could trigger only when the `!` is next to the `is`, but that only works until the user formats their code, so that really doesn't seem very useful either.

As you said some times before, we can't assume all our users have our formatter enabled, so _some_ will still have those two together.

But in my POV, this is not necessarily about them being together, it is more because of @lrhn's comment:

>  It's not useful to check for `null` just before an `is` check.

---

> (a) there will likely be too many false positives

I don't see how. You _must_ have both a `!` and an `is` next to each other, and that is not hard to find, and the `expression` before `!` must be nullable. These, are pretty easy constraints to verify (more on this at the end of this comment).

> (b) the problem isn't likely to happen often enough to be worth the extra support burden

Sure thing, I also don't have numbers, but this could also be true of https://github.com/dart-lang/sdk/issues/60518 or of the other examples by @lrhn in that issue:

> Sure. It's a classical case of "use this more specialized (and more concise and/or readable) feature", same as "use `.isEmpty` instead of `== ''`".

And this issue is the same, where the code, is not _wrong_ but it is simply unnecessary.

This can even be error-prone if the `is` is testing against a type parameter or type alias, and those change to possibly be a nullable value:

```dart
void f<T>(dynamic o) {
  if (o! is T) {}
}
```

Imagine if the caller of `f` passed in something like `int?`, then this results in a runtime error, which we could have avoided if this warning/lint existed.

---

And about @stereotype441's concern on the CL:

> The semantics of `expr! is Type` are:
> 
> - If `expr` is `null`, an exception is thrown.
> - If `expr` is not `null` and is an instance of `Type`, evaluates to `true`.
> - If `expr` is not `null` and is not an instance of `Type`, evaluates to `false`.
> 
> There's no more compact way to get those semantics than saying `expr! is Type`, so I worry that this will lead to nuisance warnings and cause some people to decide to ignore the new diagnostic.

I disagree. As above and by your explained semantics, we will throw an exception here if the user has done something wrong (null-checked a nullable value), and that is obviously acceptable, but if it can be avoided, we should aim to warn the user about it. If this check is even _unnecessary_ here to the inner block, then it makes no sense to throw an exception.

This will either be:

1. A local variable (or private field) that can be promoted: in this case, it will promote on the then branch of the if
2. A getter from an outer scope that can't be promoted, and in this case, adding the `!` will have no impact on the code inside or after the `if`

If it can be promoted to a non-nullable value, the user should do that in another branch by specifically dealing with that case (or if it is a parameter, by removing the `?` from the type annotation when possible).
